### PR TITLE
New optimizer implementation

### LINKED ISSUE #244: Optimizer does not work optimally
I have a dataset with very low optimal thresholds (lower than 0.1) , and if i to start optimizer he end his work on 0.5 with best score -118. But if i change initialGuess in optimizer.py to 0.0 and start again, i get parameter 0.1 with -93 best score.

### PR DESCRIPTION
**Pull request by scottpurdy**
fixes #244 

This PR replaces the optimization module with a new implementation that iteratively checks all possible thresholds. There are new unit tests for the scorer and the optimizer and they include a test that validates identical score calculations between the scorer and iterative scoring in the optimizer.

@BoltzmannBrain please review

### CODE REVIEW

File: nab/optimizer.py [hunk 1]
```diff
@@ -18,134 +18,303 @@
...
+import copy
+import os
+
+from nab.scorer import scaledSigmoid
+from nab import util
+
+DataRow = collections.namedtuple("DataRow", ("anomalyScore", "label"))
+"""Input format for optimizer.
+
+:param anomalyScore: the score returned by the detetor
```
Comment (BoltzmannBrain) [inline nab/optimizer.py]: detector

File: nab/optimizer.py [hunk 2]
```diff
@@ -18,134 +18,303 @@
...
+"""
+
+_WindowInfo = collections.namedtuple("_WindowInfo",
+                                     ("start", "end", "detectedAnomalies"))
+"""Private class with preprocessed window info.
+
+:param start: the file-specific index of the first point in the window
+:param end: the file-specific index of the last point in the window
+:detectedAnomalies: a list of indices of points where the detector's
+    anomaly score is higher than the current threshold. This is updated
```
Comment (BoltzmannBrain) [inline nab/optimizer.py]: higher than or equal to, no?

File: nab/optimizer.py [hunk 3]
```diff
@@ -18,134 +18,303 @@
...
+
+_DataInfo = collections.namedtuple("_DataInfo",
+                                   ("anomalyScore", "idx", "lastWindow",
+                                    "probation"))
+"""Private class with preprocessed info about a data row.
+
+:param anomalyScore: the score returned by the detector
+:param idx: the file-specific index of this row
+:param lastWindow: The _WindowInfo instance that this row is in, or the
+    last window precedign this row if this row is outside any window. None
```
Comment (BoltzmannBrain) [inline nab/optimizer.py]: preceding

File: nab/optimizer.py [hunk 4]
```diff
@@ -18,134 +18,303 @@
...
+    if no window has been seen yet in this file.
+:param probation: bool indicating whether this row is in the probation period
+"""
+
+
+
+def _processWindows(data):
+  """Process the windows into a format for easy processing.
```
Comment (BoltzmannBrain) [inline nab/optimizer.py]: better name/description would say "get" or "extract", e.g. `_extractWindowsFromData`

File: nab/optimizer.py [hunk 5]
```diff
@@ -18,134 +18,303 @@
...
+
+
+
+def _processData(data, anomalyWindows, probationaryFraction):
+  """Process the data rows into a format for easy processing.
+
+  The resulting _DataInfo instances that fall into the same window will
+  have references to the same _WindowInfo instance and thus can see which
+  other data rows have been detected so far in the
+  _WindowInfo.detectedAnomalies list.
```
Comment (BoltzmannBrain) [inline nab/optimizer.py]: "... same _WindowInfo instance, allowing us to see..."

File: nab/optimizer.py [hunk 6]
```diff
@@ -18,134 +18,303 @@
...
+  """Compute the scores for each threshold.
+
+  This is the core of the optimization. It takes the preprocessed data, sorts
+  it by anomaly score (descending), and iteratively lowers the threshold to
+  detect one more row as an anomaly. The threshold is originally set greater
+  than 1.0 and the score is set to the null detector score for the data. Then
+  the threshold is lowered to the anomaly score of each sorted row in turn.
+  Since some thresholds result in multiple rows being detected at once, we only
+  keep the score that corresponds to all of the newly-detected rows being
+  detected.
```
Comment (BoltzmannBrain) [inline nab/optimizer.py]: What does this last sentence mean? Are you referring to the (common) case where multiple rows have the same anomaly score?

File: nab/optimizer.py [hunk 7]
```diff
@@ -18,134 +18,303 @@
...
+    "fp": 0,
+    "fn": 0,
+  }
+  for dataInfo in data:
+    if (dataInfo.lastWindow is not None and
+        dataInfo.idx <= dataInfo.lastWindow.end):
+      counts["tn"] -= 1
+      counts["fn"] += 1
+
+  # Now we iteratively compute the score for every possible threshold
```
Comment (BoltzmannBrain) [inline nab/optimizer.py]: please add "We start with a threshold >1 such that no detections are made in the dataset, so the score reflects all FNs."

File: tests/unit/optimizer_test.py [hunk 8]
```diff
@@ -19,170 +19,175 @@
...
 
-"""Tests nab.optimizer for finding the local/global maxima of several 
-functions"""
-
-import math
+import datetime
 import unittest
 
-from nab.optimizer import twiddle
+import pandas
```
Comment (BoltzmannBrain) [inline tests/unit/optimizer_test.py]: alpha order imports please

File: tests/unit/optimizer_test.py [hunk 9]
```diff
@@ -19,170 +19,175 @@
...
-
-    self.assertFalse(abs(optimizedThreshold - 2.5) <= 0.00001,
-      "Optimizer found the max at the correct x = 2.5 but should not have")
-    self.assertFalse(abs(optimizedScore - 5.0625) <= 0.00001,
-      "Optimizer found the global max value of 5.0625 but should not have")
-
-
-if __name__ == '__main__':
+if __name__ == "__main__":
   unittest.main()
```
Comment (BoltzmannBrain) [inline tests/unit/optimizer_test.py]: new line at end of file

File: tests/unit/scorer_test.py [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+    # position in the window with the true positive weight. The sigmoid
+    # values are hard coded here but were calculated by running
+    # nab.scorer.scaledSigmoid.
+    fpScore = -0.9866142981514303 * COST_MATRIX["fpWeight"]
+    self.assertAlmostEqual(firstWindowScore + fpScore, score[1])
+
+
+
+if __name__ == "__main__":
+  unittest.main()
```
Comment (BoltzmannBrain) [inline tests/unit/scorer_test.py]: new line at end of file

**Comment by BoltzmannBrain** [comment]
@scottpurdy I left a few nitpicks. For the most part it looks good, but the optimizer unit tests don't run for me (stack trace below).

Please add printing to the console to show the progress; I like [tqdm](https://pypi.python.org/pypi/tqdm).

Please run this for all the detectors and push the results (thresholds file and updated scoreboard in the readme).

```
$ python tests/unit/optimizer_test.py
EEEEEEE
======================================================================
ERROR: testComputeScoreChangeFalsePositiveNoWindow (__main__.OptimizerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/unit/optimizer_test.py", line 113, in testComputeScoreChangeFalsePositiveNoWindow
    dataInfo = optimizer._DataInfo(idx=1, anomalyScore=0.1, lastWindow=None,
AttributeError: 'module' object has no attribute '_DataInfo'

======================================================================
ERROR: testComputeScoreChangeFalsePositiveWithCloseWindow (__main__.OptimizerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/unit/optimizer_test.py", line 134, in testComputeScoreChangeFalsePositiveWithCloseWindow
    windowInfo = optimizer._WindowInfo(start=3, end=5, detectedAnomalies=[])
AttributeError: 'module' object has no attribute '_WindowInfo'

======================================================================
ERROR: testComputeScoreChangeFalsePositiveWithFarWindow (__main__.OptimizerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/unit/optimizer_test.py", line 123, in testComputeScoreChangeFalsePositiveWithFarWindow
    windowInfo = optimizer._WindowInfo(start=3, end=5, detectedAnomalies=[])
AttributeError: 'module' object has no attribute '_WindowInfo'

======================================================================
ERROR: testComputeThresholdScoresCombined (__main__.OptimizerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/unit/optimizer_test.py", line 156, in testComputeThresholdScoresCombined
    results = optimizer._computeThresholdScores(_getData1() + _getData2(), 2,
AttributeError: 'module' object has no attribute '_computeThresholdScores'

======================================================================
ERROR: testComputeThresholdScoresComplex (__main__.OptimizerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/unit/optimizer_test.py", line 146, in testComputeThresholdScoresComplex
    results = optimizer._computeThresholdScores(_getData1(), 2, COST_MATRIX)
AttributeError: 'module' object has no attribute '_computeThresholdScores'

======================================================================
ERROR: testComputeThresholdScoresNoWindows (__main__.OptimizerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/unit/optimizer_test.py", line 151, in testComputeThresholdScoresNoWindows
    results = optimizer._computeThresholdScores(_getData2(), 0, COST_MATRIX)
AttributeError: 'module' object has no attribute '_computeThresholdScores'

======================================================================
ERROR: testComputeThresholdScoresPerfectDetection (__main__.OptimizerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/unit/optimizer_test.py", line 162, in testComputeThresholdScoresPerfectDetection
    data = _getData3()
  File "tests/unit/optimizer_test.py", line 94, in _getData3
    window1 = optimizer._WindowInfo(start=1, end=2, detectedAnomalies=[])
AttributeError: 'module' object has no attribute '_WindowInfo'

----------------------------------------------------------------------
Ran 7 tests in 0.001s

FAILED (errors=7)
```

File: nab/optimizer.py (hunk 6 above)
Comment (scottpurdy) [inline nab/optimizer.py]: Yes, I took another shot at the wording.

File: tests/unit/optimizer_test.py (hunk 8 above)
Comment (scottpurdy) [inline tests/unit/optimizer_test.py]: pandas is a third party library so it isn't in the same section as the built in modules (above) or the NAB imports (below)

File: tests/unit/optimizer_test.py (hunk 9 above)
Comment (scottpurdy) [inline tests/unit/optimizer_test.py]: It already has a newline at the end of the file. Note that a file ending with a newline is different than a file with a blank line at the end.

**Comment by scottpurdy** [comment]
- I'll do the progress bar in a separate PR
- I can't reproduce the test failures. `python tests/unit/optimizer_test.py` in root of repo works fine for me. Perhaps you have an old version of the code installed?
- I'll push my changes to your feedback and with updated detector results.

**Comment by smirmik** [comment]
I could be wrong, but I think in the thresholds file written wrong values "score". Because they differ from the values in scoring files. 
But...
Incredible acceleration counting !!! It is unimaginable !!! Thank you!!

**Comment by scottpurdy** [comment]
@smirnik yes this is on hold until we can track down the different in scoring. Unfortunately I haven't found a simple data sample that exhibits the scoring difference. Also note that some scores will improve slightly due to better thresholds but like you say we need to make sure the scores match between the optimizer and scorer.

**Comment by subutai** [comment]
@scottpurdy If we aren't going to address this soon, can we close this PR?  We can always reopen later.

**Comment by scottpurdy** [comment]
@subutai - yes, I think it makes more sense to switch to this for v2 when we don't have to keep exact results

**Pull request closed without merging**